### ISSUE TITLE
Deployment on serving failed

### ISSUE THREAD

**Issue by ajinkya933** [moganesyan/tensorflow_model_deployment#7]
I tried to deploy this model on google ai platform, I get this error:

```

WARNING: Using endpoint [https://ml.googleapis.com/]
Creating version (this might take a few minutes)......failed.
ERROR: (gcloud.ai-platform.versions.create) Create Version failed. Model validation failed: Outer dimension for outputs must be unknown, outer dimension of 'mrcnn_detection/Reshape_1:0' is 1 
```

Any Idea how to solve it, here's the signature def of my model:

```
signature_def['serving_default']:
  The given SavedModel SignatureDef contains the following input(s):
    inputs['input_anchors'] tensor_info:
        dtype: DT_FLOAT
        shape: (-1, -1, 4)
        name: input_anchors:0
    inputs['input_image'] tensor_info:
        dtype: DT_FLOAT
        shape: (-1, -1, -1, 3)
        name: input_image:0
    inputs['input_image_meta'] tensor_info:
        dtype: DT_FLOAT
        shape: (-1, 14)
        name: input_image_meta:0
  The given SavedModel SignatureDef contains the following output(s):
    outputs['mrcnn_detection/Reshape_1'] tensor_info:
        dtype: DT_FLOAT
        shape: (1, 100, 6)
        name: mrcnn_detection/Reshape_1:0
    outputs['mrcnn_mask/Reshape_1'] tensor_info:
        dtype: DT_FLOAT
        shape: (1, 100, 28, 28, 2)
        name: mrcnn_mask/Reshape_1:0
  Method name is: tensorflow/serving/predict

```

How can this error be fixed ?

**Comment by moganesyan** [comment]
@ajinkya933 

Hmmm its very odd that AI Platform throws this error. Our of curiosity, have you tried the standard TF-Serving way? Do you get the same error?

Anyways, it should be fixable, but you'll need to modify the original Mask r-CNN code to change the dimensions of 'mrcnn_detection/Reshape_1:0'

So for 'mrcnn_detection/Reshape_1:0' the problem seems to be here:
https://github.com/matterport/Mask_RCNN/blob/3deaec5d902d16e1daf56b62d5971d428dc920bc/mrcnn/model.py#L820


You may need to do the same for 'mrcnn_mask/Reshape_1:0' too I guess.

**Comment by ajinkya933** [comment]
Have you tried the  standard TF-Serving way? 
Yes
Do you get the same error?
No

Anyways, it should be fixable, but you'll need to modify the original Mask r-CNN code to change the dimensions of 'mrcnn_detection/Reshape_1:0'

Could you elaborate more on this?

**Comment by moganesyan** [comment]
AI Platform wants the dimension to be (-1,100,6), the Mask r-CNN code reshapes it to (batch_size, 100,6). So you should replace batch_size with -1

**Comment by ajinkya933** [comment]
So before training the model if I replace https://github.com/matterport/Mask_RCNN/blob/3deaec5d902d16e1daf56b62d5971d428dc920bc/mrcnn/model.py#L1631  

with ```batch_size=-1```. Will that fix this issue ?

**Comment by moganesyan** [comment]
the problem has nothing to do with the data generator. The issue is with the tensor dimension which is set here. 

https://github.com/matterport/Mask_RCNN/blob/3deaec5d902d16e1daf56b62d5971d428dc920bc/mrcnn/model.py#L820

**Comment by ajinkya933** [comment]
If I change https://github.com/matterport/Mask_RCNN/blob/3deaec5d902d16e1daf56b62d5971d428dc920bc/mrcnn/model.py#L820

```
        return tf.reshape(
            detections_batch,
            [self.config.BATCH_SIZE, self.config.DETECTION_MAX_INSTANCES, 6])
```

to 
```
        return tf.reshape(
            detections_batch,
            [-1, self.config.DETECTION_MAX_INSTANCES, 6])
```

does this do the trick, are there any other dependencies I need to take care of, if I do the above change ?

**Comment by moganesyan** [comment]
Yes this looks right.

Regarding the dependencies, I don't know the codebase by heart. Its up to you to try it.
